src/libmongoc/src/mongoc/mcd-azure.c

```diff
@@ -0,0 +1,116 @@
+
+void
+mcd_azure_imds_request_init (mcd_azure_imds_request *req)
+{
+   BSON_ASSERT_PARAM (req);
+   _mongoc_http_request_init (&req->req);
+   // The HTTP host of the IMDS server
+   req->req.host = "169.254.169.254";
+   // No body
+   req->req.body = "";
+   // We GET
+   req->req.method = "GET";
+   // 'Metadata: true' is required
+   req->req.extra_headers = "Metadata: true\r\n"
+                            "Accept: application/json\r\n";
+   // The default path is suitable. In the future, we may want to add query
+   // parameters to disambiguate a managed identity.
+   req->req.path = bson_strdup (DEFAULT_METADATA_PATH);
+}
+
+void
+mcd_azure_imds_request_destroy (mcd_azure_imds_request *req)
+{
+   BSON_ASSERT_PARAM (req);
+   bson_free ((void *) req->req.path);
+   *req = (mcd_azure_imds_request){0};
+}
+
+bool
+mcd_azure_access_token_try_init_from_json_str (mcd_azure_access_token *out,
+                                               const char *json,
+                                               int len,
+                                               bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT_PARAM (json);
+   bool okay = false;
+
+   if (len < 0) {
+      // Detect from a null-terminated string
+      len = strlen (json);
+   }
+
+   // Zero the output
+   *out = (mcd_azure_access_token){0};
+
+   // Parse the JSON data
+   bson_t bson;
+   if (!bson_init_from_json (&bson, json, len, error)) {
+      return false;
+   }
+
+   bson_iter_t iter;
+   // access_token
+   bool found = bson_iter_init_find (&iter, &bson, "access_token");
+   const char *const access_token =
+      !found ? NULL : bson_iter_utf8 (&iter, NULL);
+   // resource
+   found = bson_iter_init_find (&iter, &bson, "resource");
+   const char *const resource = !found ? NULL : bson_iter_utf8 (&iter, NULL);
+   // token_type
+   found = bson_iter_init_find (&iter, &bson, "token_type");
+   const char *const token_type = !found ? NULL : bson_iter_utf8 (&iter, NULL);
+   // expires_on
+   found = bson_iter_init_find (&iter, &bson, "expires_in");
+   const char *const expires_in_str =
+      !found ? NULL : bson_iter_utf8 (&iter, NULL);
+
+   if (!(access_token && resource && token_type && expires_in_str)) {
+      bson_set_error (
+         error,
+         MONGOC_ERROR_PROTOCOL_ERROR,
+         64,
+         "One or more required JSON properties are missing/invalid: data: %.*s",
+         len,
+         json);
+   } else {
+      // Set the output, duplicate each string
+      *out = (mcd_azure_access_token){
+         .access_token = bson_strdup (access_token),
+         .resource = bson_strdup (resource),
+         .token_type = bson_strdup (token_type),
+      };
+      // "expires_in" encodes the number of seconds since the issue time for
+      // which the token will be valid. strtoll() will saturate on range errors
+      // and return zero on parse errors.
+      long long s = strtoll (expires_in_str, NULL, 0);
```

Comment:
No parse failure handling (return `false`)? Immediate expiration on parse failure (expires in 0 seconds) does not seem ideal.

Reply:
I considered it, but if the parse fails and returns zero, we will still use the token exactly once, since the expiry check occurs earlier in the process. It's unfortunately not trivial to determine whether a return of zero is because of an error, or because the number is actually zero.

Reply:
I think it is worth checking over deferring the instant-expiration check to the next use of the credentials, which I suspect will often occur very soon:
```c
char *str_end;
long long s = strtoll (expires_in_str, &str_end, 0);
if (
   // Should not be an empty string.
   expires_in_str == str_end ||
   // Should have consumed entire string.
   *str_end != '\0') {
   // Parse Failure.
}
```